### PR TITLE
Add training history

### LINKED ISSUE #132: Training Run History Table
**Difficulty:** Medium
**Area:** full stack

## Description

After training a model in TensorMap, there is no record of past training runs visible in the UI. Users cannot see what models they have trained previously, what hyperparameters they used, or what the final metrics were. This information is partially stored in the database (in the `ModelBasic` table), but it is not surfaced in the frontend.

This issue adds a "Training History" table to the Training page that displays past training runs with key information: model name, creation date, epochs, optimizer, metric, and loss function. The data is read from existing database records via the model list API.

## Acceptance Criteria

- [ ] An API endpoint returns training metadata (model_name, created_on, epochs, optimizer, metric, loss, training_split) from the database
- [ ] The Training page displays a "Training History" table
- [ ] The table shows one row per saved model with key training metadata
- [ ] Rows are sorted by date (most recent first)
- [ ] An empty state message is shown when no models exist
- [ ] The table refreshes after a new model is trained and saved
- [ ] The table uses consistent styling with the rest of the page

## Resources

- `tensormap-backend/app/routers/deep_learning.py` — existing model endpoints
- `tensormap-backend/app/models/ml.py` — `ModelBasic` database model with training metadata fields
- `tensormap-frontend/src/containers/Training/Training.jsx` — training page
- `tensormap-frontend/src/services/ModelServices.js` — API service functions
- `tensormap-frontend/src/constants/Urls.js` — API URL constants
- `tensormap-frontend/src/components/ui/card.jsx` — shadcn/ui Card component

### PR DESCRIPTION
**Pull request by MohamedAli1937**
## Description

This PR adds a "Training History" table to the Training dashboard to display previously run models and their metadata so users can see a record of their past training runs.

Specifically, it includes:
- **Backend:** Updated `get_available_model_list` in `tensormap-backend/app/services/deep_learning.py` to return an ordered array of objects (sorted most recent first) containing detailed metadata (`model_name`, `created_on`, `epochs`, `optimizer`, `metric`, `loss`, `training_split`) instead of an array of strings.

- **Frontend:** Refactored `getAllModels` usage in `Training.jsx` and `Canvas.jsx` to adapt to the new object shape.
- **Frontend:** Constructed a "Training History" UI section leveraging existing Shadcn `Card` and standard HTML table components to render all training runs with date formatting and empty state handling.
- **Frontend:** Configured the component to explicitly refresh data automatically when a new run is trained or saved.

Fixes #132 

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## How Has This Been Tested?

I ran local linting on the React frontend to ensure no rules arrays were violated and all hook dependencies were properly met.

- [x] Existing tests pass
- [ ] New tests added
- [x] Manual testing

## Screenshots (if applicable)
shapes at 26-02-26 07 48 22
fixed trainig history

## **Note on Training History Issue:**

- Initially, the problem was that the `model-list` API only returned **model names**, which seemed insufficient to get full training history. The suggestion was to either create a new API (`model-history`) or extend model-list.

- After checking, we confirmed that `model_name` in `model_basic`  is **unique** (enforced by a **UNIQUE** constraint in the database), so we can safely use the names from `model-list` to fetch the full metadata from `model_basic`.

unique model_name

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review
- [x] I have added/updated documentation as needed
- [x] My changes generate no new warnings
- [x] Tests pass locally

### CODE REVIEW

**Review by ivantha — APPROVED** [review]
Lgtm 👍🏼 Please rebase.

**Comment by MohamedAli1937** [comment]
@ivantha, Commits have been squashed.

File: tensormap-backend/app/services/deep_learning.py [hunk 1]
```diff
@@ -380,15 +380,31 @@ def _numeric_dicts_to_lists(obj):
...
+            "created_on": m.created_on.isoformat() if m.created_on else None,
+            "epochs": m.epochs,
+            "optimizer": m.optimizer,
+            "metric": m.metric,
+            "loss": m.loss,
+            "training_split": m.training_split,
+        }
+        for m in models
+    ]
+
```
Comment (ivantha) [inline tensormap-backend/app/services/deep_learning.py]: **[HIGH]** Naive Datetime Serialized Without Timezone — Causes Incorrect Display in Browser

The backend serializes `created_on` using `m.created_on.isoformat()`. If `created_on` is a **naive** Python `datetime` (no `tzinfo` attached — common when the DB column lacks timezone info), the resulting ISO string will be `2024-01-15T10:30:00` with no timezone suffix.

When JavaScript's `new Date('2024-01-15T10:30:00')` parses a string **without** a timezone indicator, it treats it as **local time** on each client machine. A user in UTC+5:30 will see a time 5.5 hours ahead of a user in UTC. This is a silent, hard-to-debug data correctness bug.

The frontend then calls `.toLocaleString()`, compounding the problem since the display format becomes both timezone-incorrect and locale-variable.

**Suggestion:**
```
**Backend fix**: Ensure datetimes are timezone-aware, or explicitly mark them as UTC:
```python
# Option A: append Z to signal UTC (simple if DB stores naive UTC datetimes)
"created_on": m.created_on.isoformat() + "Z" if m.created_on else None,

# Option B: use aware datetimes (preferred long-term)
from datetime import timezone
"created_on": m.created_on.replace(tzinfo=timezone.utc).isoformat() if m.created_on else None,
```
**Frontend fix**: Use explicit UTC formatting:
```jsx
{model.created_on
  ? new Date(model.created_on).toLocaleString(undefined, { timeZone: 'UTC', timeZoneName: 'short' })
  : "-"}
```
```

File: tensormap-frontend/src/containers/Training/Training.jsx [hunk 2]
```diff
@@ -374,6 +381,7 @@ export default function Training() {
       .then((resp) => {
         if (resp.success) {
           setConfigSaved(true);
+          fetchModels();
         } else {
           logger.error("Failed to save training config:", resp.message);
```
Comment (ivantha) [inline tensormap-frontend/src/containers/Training/Training.jsx]: **[MEDIUM]** Training History Not Refreshed After Training Completes — Contradicts Stated Goal

The PR description states the component "explicitly refresh[es] data automatically when a new run is **trained** or saved" (emphasis added). However, inspecting the diff shows `fetchModels()` is only called inside the **save config** success handler (the `.then` block after saving). The WebSocket `dlResultListener` — which fires when a training run completes — does not call `fetchModels()`.

This means if a user trains a model but does not save the config, the Training History table will **not** update until the page is reloaded. This is a logic gap that contradicts the feature's stated behaviour.

**Suggestion:**
```
Add `fetchModels()` inside the DL result listener callback. Based on the socket setup pattern in the component:
```jsx
const dlResultListener = (result) => {
  // ... existing result handling logic ...
  if (result.success) {
    fetchModels(); // ← add this to refresh history after training completes
  }
};
socket.on(strings.DL_RESULT_LISTENER, dlResultListener);
```
Make sure `fetchModels` is either captured in a ref or the listener is re-registered when `fetchModels` changes, to avoid stale closure issues.
```

File: tensormap-frontend/src/containers/Training/Training.jsx [hunk 3]
```diff
@@ -805,6 +813,52 @@ export default function Training() {
...
+                    <th className="px-4 py-3 font-medium border-b">Optimizer</th>
+                    <th className="px-4 py-3 font-medium border-b">Metric</th>
+                    <th className="px-4 py-3 font-medium border-b">Loss</th>
+                    <th className="px-4 py-3 font-medium border-b">Training Split</th>
+                  </tr>
+                </thead>
+                <tbody className="divide-y">
+                  {trainingHistory.map((model, index) => (
+                    <tr key={index} className="hover:bg-muted/50">
+                      <td className="px-4 py-3 font-medium">{model.model_name}</td>
```
Comment (ivantha) [inline tensormap-frontend/src/containers/Training/Training.jsx]: **[MEDIUM]** Array Index Used as React `key` — Should Use Stable `model.id`

Table rows use `key={index}`, which is an anti-pattern in React. React uses keys to identify which items have changed, been added, or removed during reconciliation. Using an array index means:
- If the list is re-sorted or an item is deleted, React may reuse incorrect DOM nodes, causing subtle rendering bugs or stale state in child components.
- Animations, focus management, and form inputs within rows can misbehave.

The `model` objects already have a stable `id` field returned from the backend, which is the correct key to use.

**Suggestion:**
```
```jsx
// Before:
{trainingHistory.map((model, index) => (
  <tr key={index} className="hover:bg-muted/50">

// After:
{trainingHistory.map((model) => (
  <tr key={model.id} className="hover:bg-muted/50">
```
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR adds a well-scoped Training History feature by extending the model-list API with richer metadata and rendering it in a new table UI. The implementation is generally clean and follows existing patterns (Shadcn Card, useCallback, socket lifecycle). However, there are several correctness issues including a potential date/timezone display bug, a missing history refresh after training completes (contradicting the stated goal), use of array indices as React keys instead of stable IDs, and an ambiguous `training_split` value format. The API change is also technically breaking for any unconsidered consumers. None of these are showstoppers, but the timezone and training-refresh issues in particular should be resolved before merge.

File: tensormap-frontend/src/containers/Training/Training.jsx [hunk 4]
```diff
@@ -805,6 +813,52 @@ export default function Training() {
...
+                      </td>
+                    </tr>
+                  ))}
+                </tbody>
+              </table>
+            </div>
+          )}
+        </CardContent>
+      </Card>
     </div>
```
[the remark this answers is not in the document]
Comment (MohamedAli1937) [reply tensormap-frontend/src/containers/Training/Training.jsx]: Use explicit check and comment for clarity: `{model.training_split != null ? ${model.training_split}% : "-"}`

**Comment by ivantha** [comment]
Reviewed. Adds a new /model/training-history endpoint with enriched model metadata, a frontend Training History table, and an alembic migration for graph_json. Scope also includes some drive-by improvements to project service datetime/tz handling and exception breadth — acceptable refactors. CI green. Merging via rebase.
